src/target/riscv/riscv.c

```diff
@@ -1542,25 +1552,40 @@ static int riscv_examine(struct target *target)
 	/* Don't need to select dbus, since the first thing we do is read dtmcontrol. */
 
 	RISCV_INFO(info);
-	uint32_t dtmcontrol = dtmcontrol_scan(target, 0);
+	uint32_t dtmcontrol;
+	if (dtmcontrol_scan(target, 0, &dtmcontrol) != ERROR_OK || dtmcontrol == 0) {
+		LOG_TARGET_ERROR(target, "Could not read dtmcontrol. Check JTAG connectivity/board power.");
+		return ERROR_FAIL;
+	}
 	LOG_TARGET_DEBUG(target, "dtmcontrol=0x%x", dtmcontrol);
 	info->dtm_version = get_field(dtmcontrol, DTMCONTROL_VERSION);
 	LOG_TARGET_DEBUG(target, "version=0x%x", info->dtm_version);
 
+	int examination_status = ERROR_FAIL;
 	struct target_type *tt = get_target_type(target);
 	if (!tt)
-		return ERROR_FAIL;
+		goto report_examination_failure;
 
-	int result = tt->init_target(info->cmd_ctx, target);
-	if (result != ERROR_OK)
-		return result;
+	examination_status = tt->init_target(info->cmd_ctx, target);
+	if (examination_status != ERROR_OK)
+		goto report_examination_failure;
+
+	examination_status = tt->examine(target);
+	if (examination_status != ERROR_OK)
+		goto report_examination_failure;
+
+	return ERROR_OK;
 
-	return tt->examine(target);
+report_examination_failure:
+	info->dtm_version = DTM_DTMCS_VERSION_UNKNOWN;
```

Comment:
If examination of the DTM succeeds (dtmcontrol_scan()) but the later examination of the target fails, I believe we should keep the value of `info->dtm_version` because the DTM examination actually passed.

In that case, we should still be able to issue e.g. low-level DMI commands (`riscv dmi_read`, `riscv dmi_write`). I am not sure if that works today, but I believe we should not close the door on that.

Reply:
I would prefer to reset dtm_version if examine() fails. If examine() fails, it might be because the target is in some wonky state, held in reset, whatever, and the dtm version itself might be suspect. (Obviously this is not supposed to be that way, but hardware isn't always the way it's supposed to be, especially during initialization.) The added cost of detecting it again is very low

Reply:
Ok, good point, agreed.